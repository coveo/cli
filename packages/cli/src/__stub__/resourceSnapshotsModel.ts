import {ResourceSnapshotsModel} from '@coveord/platform-client';

export const getDummySnapshotModel = (
  orgId: string,
  snapshotId: string,
  {
    reports = [],
    diffGenerationReports = [],
  }: Partial<ResourceSnapshotsModel> = {}
): ResourceSnapshotsModel => ({
  id: snapshotId,
  createdBy: 'user@coveo.com',
  createdDate: 1622555047116,
  targetId: orgId,
  developerNote: 'hello',
  reports,
  diffGenerationReports,
  synchronizationReports: [],
  contentSummary: {EXTENSION: 1, FIELD: 11},
});
