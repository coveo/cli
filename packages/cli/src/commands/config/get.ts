import {Command} from '@oclif/command';
import {Config} from '../../lib/config/config';
import {ConfigRenderer} from '../../lib/config/configRenderer';
import {Trackable} from '../../lib/decorators/preconditions/trackable';

export default class Get extends Command {
  public static description = 'Display the current configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples = [
    '$ coveo config:get',
    '$ coveo config:get organization',
    '$ coveo config:get accessToken',
  ];

  @Trackable()
  public async run() {
    const {args} = this.parse(Get);
    const cfg = new Config(this.config.configDir, this.error);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }

  @Trackable()
  public async catch(err?: Error) {
    throw err;
  }
}
