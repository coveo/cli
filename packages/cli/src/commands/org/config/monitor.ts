import {ResourceSnapshotsReportModel} from '@coveord/platform-client';
import {flags} from '@oclif/command';
import {cli} from 'cli-ux';
import {
  IsAuthenticated,
  Preconditions,
} from '../../../lib/decorators/preconditions';
import {ReportViewerStyles} from '../../../lib/snapshot/reportViewer/reportViewerStyles';
import {Snapshot, waitUntilDoneOptions} from '../../../lib/snapshot/snapshot';
import {SnapshotFactory} from '../../../lib/snapshot/snapshotFactory';
import SnapshotBase from './orgConfigBase';

export default class Monitor extends SnapshotBase {
  public static description = 'Monitor a Snapshot operation';

  public static flags = {
    target: flags.string({
      char: 't',
      description:
        'The unique identifier of the organization containing the snapshot. If not specified, the organization you are connected to will be used.',
      helpValue: 'destinationorganizationg7dg3gd',
      required: false,
    }),
  };

  public static args = [
    {
      name: 'snapshotId',
      description: 'The unique identifier of the target snapshot.',
      required: true,
    },
  ];

  public static hidden = true;

  @Preconditions(IsAuthenticated())
  public async run() {
    const snapshot = await this.getSnapshot();

    this.printHeader(snapshot.id);

    await this.monitorSnapshot(snapshot);
  }

  private async monitorSnapshot(snapshot: Snapshot) {
    cli.action.start(
      `Operation: ${this.getReportType(snapshot.latestReport)}`,
      this.getReportStatus(snapshot.latestReport)
    );

    // TODO: revisit with a progress bar once the response contains the remaining resources to process
    const iteratee = (report: ResourceSnapshotsReportModel) =>
      this.refresh(report);
    await snapshot.waitUntilDone(this.waitOptions, iteratee);

    cli.action.stop(this.getReportStatus(snapshot.latestReport));
  }

  private printHeader(snapshotId: string) {
    const header = ReportViewerStyles.header(
      `\nMonitoring snapshot ${snapshotId}:`
    );
    cli.log(header);
  }

  private prettyPrint(str: string): string {
    const capitalized = str.charAt(0) + str.slice(1).toLowerCase();
    return capitalized.replace(/_/g, ' ');
  }

  private getReportType(report: ResourceSnapshotsReportModel) {
    const type = this.prettyPrint(report.type);
    return type;
  }

  private getReportStatus(report: ResourceSnapshotsReportModel) {
    const status = this.prettyPrint(report.status);
    return status;
  }

  private refresh(report: ResourceSnapshotsReportModel) {
    const status = this.getReportStatus(report);
    cli.action.status = status;
  }

  private async getSnapshot(): Promise<Snapshot> {
    const {args} = this.parse(Monitor);
    const snapshotId = args.snapshotId;
    const target = await this.getTargetOrg();

    return SnapshotFactory.createFromExistingSnapshot(snapshotId, target);
  }

  private get waitOptions(): waitUntilDoneOptions {
    return {
      waitOptions: {
        numOfAttempts: Infinity,
        delayFirstAttempt: false,
        maxDelay: 10e3,
      },
    };
  }
}
