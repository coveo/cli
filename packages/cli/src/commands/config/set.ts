import {Command, flags} from '@oclif/command';
import {Config} from '../../lib/config/config';
import {AuthenticatedClient} from '../../lib/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '../../lib/decorators/preconditions';
import {PlatformEnvironment} from '../../lib/platform/environment';
import {withEnvironment, withRegion} from '../../lib/flags/platformCommonFlags';
import {Region} from '@coveord/platform-client';
import {Trackable} from '../../lib/decorators/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';

export default class Set extends Command {
  public static description = 'Modify the current configuration.';

  public static flags = {
    ...withRegion(false),
    ...withEnvironment(false),
    organization: flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
    analytics: flags.string({
      char: 'a',
      options: ['y', 'n'],
      description: 'Whether to enable analytics and telemetry tracking.',
    }),
  };

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = this.parse(Set);
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir, this.error);
    if (flags.environment) {
      cfg.set('environment', flags.environment as PlatformEnvironment);
    }
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }
    if (flags.region) {
      cfg.set('region', flags.region as Region);
    }
    if (flags.analytics) {
      cfg.set('analyticsEnabled', flags.analytics === 'y');
    }
  }

  @Trackable()
  public async catch(err?: Error) {
    throw err;
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
