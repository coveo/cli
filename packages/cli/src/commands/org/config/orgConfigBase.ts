import {Command, flags} from '@oclif/command';
import {cli} from 'cli-ux';
import {ReadStream} from 'fs';
import {dedent} from 'ts-dedent';
import {cwd} from 'process';
import {Config} from '../../../lib/config/config';
import {Project} from '../../../lib/project/project';
import {SnapshotFactory} from '../../../lib/snapshot/snapshotFactory';
import {
  snapshotSynchronizationUrl,
  snapshotUrl,
  SnapshotUrlOptionsArgs,
} from '../../../lib/platform/url';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {red, green} from 'chalk';
import {normalize} from 'path';
import {SnapshotReporter} from '../../../lib/snapshot/snapshotReporter';

export interface CustomFile extends ReadStream {
  type?: string;
}

export default abstract class SnapshotBase extends Command {
  public static description = 'Create and validate a snapshot';

  public static flags = {
    target: flags.string({
      char: 't',
      description:
        'The unique identifier of the organization where to send the changes. If not specified, the organization you are connected to will be used.',
      helpValue: 'destinationorganizationg7dg3gd',
      required: false,
    }),
    deleteMissingResources: flags.boolean({
      char: 'd',
      description: 'Whether or not to show resources to delete',
      default: false,
      required: false,
    }),
  };

  public static hidden = true;

  protected async dryRun() {
    const project = new Project(normalize(this.projectPath));

    cli.action.start('Creating snapshot');
    const snapshot = await this.createSnapshotFromProject(project);

    cli.action.start('Validating snapshot');
    const reporter = await this.validateSnapshot(snapshot);

    cli.action.stop(reporter.isSuccessReport() ? green('âœ”') : red.bold('!'));
    return {reporter, snapshot, project};
  }

  protected async createSnapshotFromProject(
    project: Project
  ): Promise<Snapshot> {
    const pathToZip = await project.compressResources();
    const targetOrg = await this.getTargetOrg();

    return await SnapshotFactory.createFromZip(pathToZip, targetOrg);
  }

  protected async validateSnapshot(
    snapshot: Snapshot
  ): Promise<SnapshotReporter> {
    const reporter = await snapshot.validate(this.flags.deleteMissingResources);

    if (!reporter.isSuccessReport()) {
      await this.handleReportWithErrors(snapshot);
    }

    return reporter;
  }

  protected async getTargetOrg() {
    if (this.flags.target) {
      return this.flags.target;
    }
    const cfg = await this.configuration.get();
    return cfg.organization;
  }

  protected get projectPath() {
    return cwd();
  }

  protected async handleReportWithErrors(snapshot: Snapshot) {
    // TODO: CDX-362: handle invalid snapshot cases
    const pathToReport = snapshot.saveDetailedReport(this.projectPath);
    const report = snapshot.latestReport;

    if (snapshot.requiresSynchronization()) {
      const synchronizationPlanUrl = await this.getSynchronizationPage(
        snapshot
      );
      this.warn(
        dedent`
        Some conflicts were detected while comparing changes between the snapshot and the target organization.
        Click on the URL below to synchronize your snapshot with your organization before running the command again.
        ${synchronizationPlanUrl}
        `
      );
      return;
    }

    const snapshotUrl = await this.getSnapshotPage(snapshot);

    this.error(
      dedent`Invalid snapshot - ${report.resultCode}.
      Detailed report saved at ${pathToReport}.

      You can also use this link to view the snapshot in the Coveo Admin Console
      ${snapshotUrl}`
    );
  }

  private get flags() {
    const {flags} = this.parse(this.ctor as typeof SnapshotBase);
    return flags;
  }

  private get configuration() {
    return new Config(this.config.configDir, this.error);
  }

  private async getSnapshotPage(snapshot: Snapshot) {
    const options = await this.getSnapshotUrlOptions(snapshot);
    return snapshotUrl(options);
  }

  private async getSynchronizationPage(snapshot: Snapshot) {
    const options = await this.getSnapshotUrlOptions(snapshot);
    return snapshotSynchronizationUrl(options);
  }

  private async getSnapshotUrlOptions(
    snapshot: Snapshot
  ): Promise<SnapshotUrlOptionsArgs> {
    const {environment} = await this.configuration.get();
    return {
      environment,
      targetOrgId: snapshot.targetId,
      snapshotId: snapshot.id,
    };
  }
}
