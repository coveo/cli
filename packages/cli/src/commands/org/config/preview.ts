import {DryRunOptions} from '@coveord/platform-client';
import {Command, flags} from '@oclif/command';
import {cwd} from 'process';
import {
  buildAnalyticsFailureHook,
  buildAnalyticsSuccessHook,
} from '../../../hooks/analytics/analytics';
import {Config} from '../../../lib/config/config';
import {
  IsAuthenticated,
  Preconditions,
} from '../../../lib/decorators/preconditions';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {
  displayInvalidSnapshotError,
  displaySnapshotSynchronizationWarning,
  dryRun,
  getTargetOrg,
  handleSnapshotError,
} from '../../../lib/snapshot/snapshotCommon';

export default class Preview extends Command {
  public static description = 'Preview resource updates';

  public static flags = {
    target: flags.string({
      char: 't',
      description:
        'The unique identifier of the organization where to send the changes. If not specified, the organization you are connected to will be used.',
      helpValue: 'destinationorganizationg7dg3gd',
      required: false,
    }),
    showMissingResources: flags.boolean({
      char: 'd',
      description: 'Whether or not preview missing resources',
      default: false,
      required: false,
    }),
  };

  public static hidden = true;

  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = this.parse(Preview);
    const target = await getTargetOrg(this.configuration, flags.target);
    const options: DryRunOptions = {
      deleteMissingResources: flags.showMissingResources,
    };
    const {reporter, snapshot, project} = await dryRun(
      target,
      this.projectPath,
      options
    );

    await snapshot.preview();

    if (reporter.isSuccessReport()) {
      await snapshot.delete();
    } else {
      await this.handleReportWithErrors(snapshot);
    }

    project.deleteTemporaryZipFile();

    this.config.runHook('analytics', buildAnalyticsSuccessHook(this, flags));
  }

  public async catch(err?: Error) {
    const {flags} = this.parse(Preview);
    handleSnapshotError(err);
    await this.config.runHook(
      'analytics',
      buildAnalyticsFailureHook(this, flags, err)
    );
  }

  private async handleReportWithErrors(snapshot: Snapshot) {
    // TODO: CDX-362: handle invalid snapshot cases
    const cfg = await this.configuration.get();

    if (snapshot.requiresSynchronization()) {
      displaySnapshotSynchronizationWarning(snapshot, cfg);
      return;
    }

    displayInvalidSnapshotError(snapshot, cfg, this.projectPath);
  }

  private get configuration() {
    return new Config(this.config.configDir, this.error);
  }

  private get projectPath() {
    return cwd();
  }
}
