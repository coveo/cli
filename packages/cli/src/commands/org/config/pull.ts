import {ResourceSnapshotType} from '@coveord/platform-client';
import {flags, Command} from '@oclif/command';
import {IOptionFlag} from '@oclif/command/lib/flags';
import {blueBright} from 'chalk';
import {cli} from 'cli-ux';
import {cwd} from 'process';
import dedent from 'ts-dedent';
import {buildAnalyticsFailureHook} from '../../../hooks/analytics/analytics';
import {Config} from '../../../lib/config/config';
import {
  IsAuthenticated,
  Preconditions,
} from '../../../lib/decorators/preconditions';
import {SnapshotOperationTimeoutError} from '../../../lib/errors';
import {Project} from '../../../lib/project/project';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {
  getTargetOrg,
  handleSnapshotError,
} from '../../../lib/snapshot/snapshotCommon';
import {SnapshotFactory} from '../../../lib/snapshot/snapshotFactory';

export default class Pull extends Command {
  public static description = 'Pull resources from an organization';

  public static flags = {
    target: flags.string({
      char: 't',
      helpValue: 'destinationorganizationg7dg3gd',
      required: false,
      description:
        'The unique identifier of the organization from which to pull the resources. If not specified, the organization you are connected to will be used.',
    }),
    resourceTypes: flags.string({
      char: 'r',
      helpValue: 'type1 type2',
      options: Object.keys(ResourceSnapshotType),
      default: Object.keys(ResourceSnapshotType),
      multiple: true,
      description: 'The resources types to pull from the organization.',
    }) as IOptionFlag<(keyof typeof ResourceSnapshotType)[]>,
    snapshotId: flags.string({
      char: 's',
      exclusive: ['resourceTypes'],
      description:
        'The unique identifier of the snapshot to pull. If not specified, a new snapshot will be created.',
    }),
  };

  public static hidden = true;

  @Preconditions(IsAuthenticated())
  public async run() {
    cli.action.start('Creating Snapshot');
    const snapshot = await this.getSnapshot();

    cli.action.start('Updating project with Snapshot');
    await this.refreshProject(snapshot);

    await snapshot.delete();
    cli.action.stop('Project updated');
  }

  public async catch(err?: Error) {
    const {flags} = this.parse(Pull);
    await this.config.runHook(
      'analytics',
      buildAnalyticsFailureHook(this, flags, err)
    );
    handleSnapshotError(err);
    this.displayAdditionalErrorMessage(err);
  }

  private displayAdditionalErrorMessage(err?: Error) {
    if (err instanceof SnapshotOperationTimeoutError) {
      const snapshot = err.snapshot;
      cli.info(
        dedent`

          Once the snapshot is created, you can pull it with the following command:

            ${blueBright`coveo org:config:pull ${snapshot.id}`}

            `
      );
    }
  }

  private async refreshProject(snapshot: Snapshot) {
    const project = new Project(this.projectPath);
    const snapshotBlob = await snapshot.download();
    await project.refresh(snapshotBlob);
  }

  private async getSnapshot() {
    const {flags} = this.parse(Pull);
    const target = await getTargetOrg(this.configuration, flags.target);
    if (flags.snapshotId) {
      return SnapshotFactory.createFromExistingSnapshot(
        flags.snapshotId,
        target
      );
    }
    return SnapshotFactory.createFromOrg(
      this.ResourceSnapshotTypesToExport,
      target
    );
  }

  private get configuration() {
    return new Config(this.config.configDir, this.error);
  }

  private get ResourceSnapshotTypesToExport() {
    const {flags} = this.parse(Pull);
    return flags.resourceTypes.map((type) => ResourceSnapshotType[type]);
  }

  private get projectPath() {
    return cwd();
  }
}
