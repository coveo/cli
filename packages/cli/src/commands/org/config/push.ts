import {cli} from 'cli-ux';
import {flags, Command} from '@oclif/command';
import {
  IsAuthenticated,
  Preconditions,
} from '../../../lib/decorators/preconditions';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {red, green, bold} from 'chalk';
import {SnapshotReporter} from '../../../lib/snapshot/snapshotReporter';
import {
  displayInvalidSnapshotError,
  displaySnapshotSynchronizationWarning,
  dryRun,
  DryRunOptions,
  getTargetOrg,
  handleSnapshotError,
} from '../../../lib/snapshot/snapshotCommon';
import {Config} from '../../../lib/config/config';
import {cwd} from 'process';
import {
  buildAnalyticsFailureHook,
  buildAnalyticsSuccessHook,
} from '../../../hooks/analytics/analytics';

export default class Push extends Command {
  public static description =
    'Preview, validate and deploy your changes to the destination org';

  public static flags = {
    target: flags.string({
      char: 't',
      description:
        'The unique identifier of the organization where to send the changes. If not specified, the organization you are connected to will be used.',
      helpValue: 'destinationorganizationg7dg3gd',
      required: false,
    }),
    deleteMissingResources: flags.boolean({
      char: 'd',
      description: 'Whether or not to delete missing resources',
      default: false,
      required: false,
    }),
    skipPreview: flags.boolean({
      char: 's',
      description:
        'Do not preview changes before applying them to the organization',
      default: false,
      required: false,
    }),
  };

  public static hidden = true;

  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = this.parse(Push);
    const target = await getTargetOrg(this.configuration, flags.target);
    const options: DryRunOptions = {
      deleteMissingResources: flags.deleteMissingResources,
    };
    const {reporter, snapshot, project} = await dryRun(
      target,
      this.projectPath,
      options
    );

    if (!flags.skipPreview) {
      await snapshot.preview();
    }

    if (reporter.isSuccessReport()) {
      await this.handleValidReport(reporter, snapshot);
      await snapshot.delete();
    } else {
      await this.handleReportWithErrors(snapshot);
    }

    project.deleteTemporaryZipFile();

    this.config.runHook('analytics', buildAnalyticsSuccessHook(this, flags));
  }

  public async catch(err?: Error) {
    const {flags} = this.parse(Push);
    handleSnapshotError(err);
    await this.config.runHook(
      'analytics',
      buildAnalyticsFailureHook(this, flags, err)
    );
  }

  private async handleValidReport(
    reporter: SnapshotReporter,
    snapshot: Snapshot
  ) {
    if (!reporter.hasChangedResources()) {
      return;
    }

    const {flags} = this.parse(Push);
    const canBeApplied = flags.skipPreview || (await this.askForConfirmation());

    if (canBeApplied) {
      await this.applySnapshot(snapshot);
    }
  }

  private async askForConfirmation() {
    const {flags} = this.parse(Push);
    const target = await getTargetOrg(this.configuration, flags.target);
    const canBeApplied = await cli.confirm(
      `\nWould you like to apply these changes to the org ${bold(
        target
      )}? (y/n)`
    );
    return canBeApplied;
  }

  private async applySnapshot(snapshot: Snapshot) {
    cli.action.start('Applying snapshot');
    const {flags} = this.parse(Push);
    const reporter = await snapshot.apply(flags.deleteMissingResources);
    const success = reporter.isSuccessReport();

    if (!success) {
      await this.handleReportWithErrors(snapshot);
    }

    cli.action.stop(success ? green('âœ”') : red.bold('!'));
  }

  private async handleReportWithErrors(snapshot: Snapshot) {
    const cfg = await this.configuration.get();

    if (snapshot.requiresSynchronization()) {
      displaySnapshotSynchronizationWarning(snapshot, cfg);
      return;
    }

    displayInvalidSnapshotError(snapshot, cfg, this.projectPath);
  }

  private get configuration() {
    return new Config(this.config.configDir, this.error);
  }

  private get projectPath() {
    return cwd();
  }
}
