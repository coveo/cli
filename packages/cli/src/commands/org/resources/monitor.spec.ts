jest.mock('../../../lib/config/config');
jest.mock('../../../hooks/analytics/analytics');

jest.mock('../../../lib/platform/authenticatedClient');
jest.mock('../../../lib/snapshot/snapshotReporter');
jest.mock('../../../lib/snapshot/snapshotFactory');

import {Config} from '../../../lib/config/config';
import {ResourceSnapshotsReportType} from '@coveord/platform-client';
import {test} from '@oclif/test';
import {getDummySnapshotModel} from '../../../__stub__/resourceSnapshotsModel';
import {getSuccessReport} from '../../../__stub__/resourceSnapshotsReportModel';
import {SnapshotReporter} from '../../../lib/snapshot/snapshotReporter';
import {SnapshotReportStatus} from '../../../lib/snapshot/reportPreviewer/reportPreviewerDataModels';
import {SnapshotFactory} from '../../../lib/snapshot/snapshotFactory';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {formatCliLog} from '../../../__test__/jestSnapshotUtils';

const mockedSnapshotFactory = jest.mocked(SnapshotFactory, true);
const mockedConfig = jest.mocked(Config);
const mockedConfigGet = jest.fn();
const mockedGetSnapshot = jest.fn();
const mockedWaitUntilDone = jest.fn();
const mockedValidateSnapshot = jest.fn();
const mockedPreviewSnapshot = jest.fn();
const mockedDeleteSnapshot = jest.fn();
const mockedSaveDetailedReport = jest.fn();
const mockedAreResourcesInError = jest.fn();
const mockedLatestReport = jest.fn();

const mockedSnapshotReporter = jest.mocked(SnapshotReporter, true);

const mockSnapshotFactory = () => {
  mockedSnapshotFactory.createFromExistingSnapshot.mockReturnValue(
    Promise.resolve({
      validate: mockedValidateSnapshot,
      preview: mockedPreviewSnapshot,
      delete: mockedDeleteSnapshot,
      saveDetailedReport: mockedSaveDetailedReport,
      areResourcesInError: mockedAreResourcesInError,
      latestReport: mockedLatestReport,
      waitUntilDone: mockedWaitUntilDone,
      id: 'banana-snapshot',
      targetId: 'potato-org',
    } as unknown as Snapshot)
  );
};

const doMockConfig = () => {
  mockedConfigGet.mockReturnValue({
    region: 'us',
    organization: 'default-org',
    environment: 'prod',
  });

  mockedConfig.mockImplementation(
    () =>
      ({
        get: mockedConfigGet,
      } as unknown as Config)
  );
};

const doMockSnapshotReporter = (snapshotReportStatus: SnapshotReportStatus) => {
  let handlerToCall: (
    this: SnapshotReporter
  ) => void | Promise<void> = () => {};
  mockedSnapshotReporter.prototype.setReportHandler.mockImplementation(
    function (
      this: SnapshotReporter,
      status: SnapshotReportStatus,
      handler: (this: SnapshotReporter) => void | Promise<void>
    ) {
      if (status === snapshotReportStatus) {
        handlerToCall = handler;
      }
      return this;
    }
  );
  mockedSnapshotReporter.prototype.handleReport.mockImplementation(
    async function (this: SnapshotReporter) {
      await handlerToCall.apply(this);
    }
  );
};

describe('org:resources:monitor', () => {
  beforeAll(() => {
    doMockConfig();
    mockSnapshotFactory();
    doMockSnapshotReporter(SnapshotReportStatus.SUCCESS);
    mockedGetSnapshot.mockResolvedValue(
      getDummySnapshotModel('default-org', 'my-snapshot', [
        getSuccessReport('my-snapshot', ResourceSnapshotsReportType.Apply),
      ])
    );
  });

  test
    .stdout()
    .stderr()
    .command(['org:resources:monitor'])
    .catch(/Missing 1 required arg/)
    .it('requires snapshotId name argument');

  test
    .stdout()
    .stderr()
    .command(['org:resources:monitor', 'my-snapshot'])
    .it('should work with default connected org', () => {
      expect(
        mockedSnapshotFactory.createFromExistingSnapshot
      ).toHaveBeenCalledWith('my-snapshot', 'default-org');
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:monitor', 'other-snapshot', '-o', 'different-org'])
    .it('should work with default connected org', () => {
      expect(
        mockedSnapshotFactory.createFromExistingSnapshot
      ).toHaveBeenCalledWith('other-snapshot', 'different-org');
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:monitor', 'my-snapshot'])
    .it(
      'should create two differents reporters, one before the action is complete and another afterward',
      () => {
        const firstReport = mockedLatestReport.mock.instances[0];
        const lastReport = mockedLatestReport.mock.instances[1];
        const firstReportParameterId =
          mockedSnapshotReporter.mock.calls[0][0].id;
        const lastReportParameterId =
          mockedSnapshotReporter.mock.calls[
            mockedSnapshotReporter.mock.calls.length - 1
          ][0].id;
        // First call is the one used prior to the refresh loop
        expect(mockedSnapshotReporter).toBeCalledTimes(2);
        expect(firstReportParameterId).toBe(firstReport);
        // Last call is used when the refresh loop is complete.
        expect(lastReportParameterId).toBe(lastReport);
      }
    );

  describe.each([
    {
      describeName: 'when the operation fails',
      reporterMockedStatus: SnapshotReportStatus.ERROR,
    },
    {
      describeName: 'when the operation succeed',
      reporterMockedStatus: SnapshotReportStatus.SUCCESS,
    },
    {
      describeName: 'when the operations conclude with missing vault entries',
      reporterMockedStatus: SnapshotReportStatus.MISSING_VAULT_ENTRIES,
    },
    {
      describeName: 'when the operation conclude with no changes',
      reporterMockedStatus: SnapshotReportStatus.NO_CHANGES,
    },
  ])('$describeName', ({reporterMockedStatus}) => {
    beforeAll(() => {
      doMockSnapshotReporter(reporterMockedStatus);
    });

    test
      .stdout()
      .stderr()
      .command(['org:resources:monitor', 'my-snapshot'])
      .it('should output the same thing', (ctx) => {
        expect(formatCliLog(ctx.stdout)).toMatchSnapshot();
        expect(formatCliLog(ctx.stderr)).toMatchSnapshot();
      });
  });
});
