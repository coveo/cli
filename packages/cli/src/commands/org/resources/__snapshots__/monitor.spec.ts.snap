// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org:resources:monitor when the operation conclude with no changes should output the same thing 1`] = `
"
"
`;

exports[`org:resources:monitor when the operation conclude with no changes should output the same thing 2`] = `
"Monitoring snapshot my-snapshot...
Operation undefined... ✔
"
`;

exports[`org:resources:monitor when the operation fails should output the same thing 1`] = `
"
undefined
"
`;

exports[`org:resources:monitor when the operation fails should output the same thing 2`] = `
"Monitoring snapshot my-snapshot...
Operation undefined... ✔
"
`;

exports[`org:resources:monitor when the operation succeed should output the same thing 1`] = `
"
"
`;

exports[`org:resources:monitor when the operation succeed should output the same thing 2`] = `
"Monitoring snapshot my-snapshot...
Operation undefined... ✔
"
`;

exports[`org:resources:monitor when the operations conclude with missing vault entries should output the same thing 1`] = `
"
"
`;

exports[`org:resources:monitor when the operations conclude with missing vault entries should output the same thing 2`] = `
"Monitoring snapshot my-snapshot...
Operation undefined... ✔
"
`;
